testsuite/tests/typing-layouts-or-null/runtime.ml

```diff
@@ -0,0 +1,150 @@
+(* TEST
+   flags = "-extension layouts_alpha";
+   runtime5;
+*)
+
+let x = Null
+
+let () =
+  match x with
+  | Null -> ()
+  | This _ -> assert false
+;;
+
+let y = This 3
+
+let () =
+  match y with
+  | This 3 -> ()
+  | _ -> assert false
+;;
+
+
+external int_as_pointer : int -> int or_null = "%int_as_pointer"
+
+let n = int_as_pointer 0
+
+let () =
+  match n with
+  | Null -> ()
+  | _ -> assert false
+;;
+
+external int_as_int : int -> int or_null = "%identity"
+
+let m = int_as_int 5
+
+let () =
+  match m with
+  | This 5 -> ()
+  | This _ -> assert false
+  | Null -> assert false
+;;
+
+let x = (Null, This "bar")
+
+let () =
+  match x with
+  | Null, This "foo" -> assert false
+  | Null, This "bar" -> ()
+  | _, This "bar" -> assert false
+  | Null, _ -> assert false
+  | _, _ -> assert false
+;;
+
+let y a = fun () -> This a
+
+let d = y 5
+
+let () =
+  match d () with
+  | This 5 -> ()
+  | _ -> assert false
+;;
+
+external to_bytes : ('a : value_or_null) . 'a -> int list -> bytes = "caml_output_value_to_bytes"
+
+external from_bytes_unsafe : ('a : value_or_null) . bytes -> int -> 'a = "caml_input_value_from_bytes"
+
+let z = to_bytes (This "foo") []
+
+let () =
+  match from_bytes_unsafe z 0 with
+    | This "foo" -> ()
+    | This _ -> assert false
+    | Null -> assert false
+;;
+
+let w = to_bytes Null []
+
+let () =
+  match from_bytes_unsafe w 0 with
+    | Null -> ()
+    | This _ -> assert false
+;;
+
+external evil : 'a or_null -> 'a = "%identity"
+
+let e = This (evil Null)
+
+let () =
+  match e with
+  | Null -> ()
+  | This _ -> assert false
+;;
```

Comment:
> In particular, Flambda2 should be allowed to make slightly different versions of this test fail.
Would this be true if the test uses `%obj_magic` instead of `%identity`?